File: notes/graph/dfs_with_stack.py
class Graph:
    def __init__(self, vertices):
        self.vertices = vertices
        # Adjacency list
        self.adjacency_list = [[] for _ in range(vertices)]

    def add_edge(self, source, destination):
        self.adjacency_list[source].append(destination)
        # For an undirected graph
        self.adjacency_list[destination].append(source)

    def dfs(self, starting_vertex):
        visited = [False] * self.vertices
        stack = []

        stack.append(starting_vertex)

        while stack:
            current_vertex = stack.pop()

            if not visited[current_vertex]:
                # Process the current node
                print(current_vertex, end=" ")
                visited[current_vertex] = True

            # Push all unvisited neighbors onto the stack
            for neighbor in self.adjacency_list[current_vertex]:
                if not visited[neighbor]:
                    stack.append(neighbor)

# V = number of vertices, E = number of edges
# Time complexity: O(V + E)
# __init__(): O(V)
# dfs(): O(E)
# Space complexity: O(V)
# __init__(): O(V)
# dfs(): O(V)

File: notes/graph/dfs_with_stack.cpp
#include <iostream>
#include <stack>
#include <vector>

using namespace std;

class Graph {
  private:
    int vertices;
    vector<vector<int>> adjacencyList;

  public:
    Graph(int vertices) : vertices(vertices), adjacencyList(vertices) {}

    void addEdge(int source, int destination) {
        adjacencyList[source].push_back(destination);
        adjacencyList[destination].push_back(source);
    }

    void dfs(int startingVertex) {
        vector<bool> visited(vertices, false);
        stack<int> myStack;
        myStack.push(startingVertex);

        while (!myStack.empty()) {
            int curr = myStack.top();
            myStack.pop();

            if (!visited[curr]) {
                cout << curr << " ";
                visited[curr] = true;
            }

            for (auto neighbor : adjacencyList[curr]) {
                if (!visited[neighbor]) {
                    myStack.push(neighbor);
                }
            }
        }
    }
};

// V = number of vertices, E = number of edges
// Time complexity: O(V + E)
// constructor(): O(V)
// dfs(): O(E)
// Space complexity: O(V)
// constructor(): O(V)
// dfs(): O(V)